### PR TITLE
cleanup(generator): refactor GetEffectiveServiceName

### PR DESCRIPTION
**Pull request by alevenberg**
Closes https://github.com/googleapis/google-cloud-cpp/issues/12935

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12945)

### CODE REVIEW

**Comment by alevenberg** [comment]
I'm not sure passing the argument `name` as a `absl::string_view name` makes it more readable. But maybe the efficiency is a good enough reason to change it.

**Comment by alevenberg** [comment]
This is why I used a colon I think

https://github.com/alevenberg/google-cloud-cpp/blob/dbbd0c440dcaa124c049ccab85ef321e4954e3f5/generator/standalone_main.cc#L338-L342

I'm not sure if it will break or not. Looking into...

File: generator/internal/descriptor_utils.cc [hunk 1]
```diff
@@ -634,20 +634,19 @@ bool CheckParameterCommentSubstitutions() {
...
-  if (service_name_mappings == vars.end()) return name;
-
-  std::vector<std::string> mappings =
-      absl::StrSplit(service_name_mappings->second, ',');
-  for (auto& mapping : mappings) {
-    std::vector<std::string> kv = absl::StrSplit(mapping, ':');
-    if (kv.size() != 2) continue;
-    if (kv[0] == name) return kv[1];
+  if (service_name_mappings == vars.end())
+    return std::string(name.data(), name.size());
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:642]: Think about coalescing this with the final return by reversing the `if` guard.

File: generator/internal/descriptor_utils.cc [hunk 2]
```diff
@@ -634,20 +634,19 @@ bool CheckParameterCommentSubstitutions() {
...
+  std::map<absl::string_view, std::string> map;
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:643]: The range of the map can also be a `string_view`.  It is only when we have a successful lookup that we have to convert anything to a `string`.

File: generator/internal/descriptor_utils.cc [hunk 3]
```diff
@@ -634,20 +634,19 @@ bool CheckParameterCommentSubstitutions() {
...
+  for (absl::string_view arg :
+       absl::StrSplit(service_name_mappings->second, ',')) {
+    map.insert(absl::StrSplit(arg, absl::MaxSplits('=', 1)));
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:646]: Doesn't the change to `=` here need to be in the same PR that changes the generation of the string?

**Comment by devbww** [comment]
> I'm not sure passing the argument `name` as a `absl::string_view name` makes it more readable. But maybe the efficiency is a good enough reason to change it.

Yes ... hopefully the readability is exactly the same as the lifetimes of the `string_view`s are clear.

File: generator/internal/descriptor_utils.cc (hunk 1 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Done

File: generator/internal/descriptor_utils.cc (hunk 2 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Done

File: generator/internal/descriptor_utils.cc (hunk 3 above)
Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Yes, if I was already using the service_name_mappings in the config, but I have not commited any of those changes yet.

File: generator/internal/descriptor_utils.cc [hunk 4]
```diff
@@ -634,20 +634,22 @@ bool CheckParameterCommentSubstitutions() {
...
+  if (service_name_mappings != vars.end()) {
+    std::map<absl::string_view, absl::string_view> map;
+    for (absl::string_view arg :
+         absl::StrSplit(service_name_mappings->second, ',')) {
+      map.insert(absl::StrSplit(arg, absl::MaxSplits('=', 1)));
+    }
+    if (map.find(name) != map.end()) {
+      return std::string(map.find(name)->second.data(),
+                         map.find(name)->second.size());
+    }
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:650]: Saving the result of the first `find()` would be preferable to running the same lookup three times.

Comment (alevenberg) [reply generator/internal/descriptor_utils.cc]: Done.

File: generator/internal/descriptor_utils.cc [hunk 5]
```diff
@@ -634,330 +634,337 @@ bool CheckParameterCommentSubstitutions() {
...
+    std::map<absl::string_view, absl::string_view> map;
+    for (absl::string_view arg :
+         absl::StrSplit(service_name_mappings->second, ',')) {
+      map.insert(absl::StrSplit(arg, absl::MaxSplits('=', 1)));
+    }
+    auto result = map.find(name);
+    if (result != map.end()) {
+      return std::string(result->second.data(), result->second.size());
+    }
+    return std::string(name.data(), name.size());
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:651]: In the future, when there is explicit `string_view` to `string` conversion, these will be a little nicer.

**Review by devbww — APPROVED** [review]

**Pull request merged**